src/runner.js

```diff
@@ -28,12 +28,96 @@ const debugResponse = _debug('remote-s3:Runner:res');
  * actually transmitted matched a pre-computed size and hash
  */
 class Runner {
-  constructor(agent, agentOpts) {
-    this.agent = agent || new https.Agent(agentOpts || {}); 
+  constructor(opts) {
+    opts = runSchema(opts || {}, Joi.object().keys({
+      agent: Joi.object(),
+      agentOpts: Joi.object(),
+      maxRetries: Joi.number().min(0).max(10).default(5),
+    }).without('agent', 'agentOpts').optionalKeys(['agent', 'agentOpts']));
+
+    this.agent = opts.agent || new https.Agent(opts.agentOpts || {}); 
+
+    this.maxRetries = opts.maxRetries;
+
   }
 
+  /**
+   * Run a request and retry if a retryable error occurs.  Only 500-series
+   * responses are considered retriable, so any other status code will return
+   * with its status.
+   *
+   * http://docs.aws.amazon.com/general/latest/gr/api-retries.html
+   */
   async run(opts) {
-    return this.runOnce(opts);
+    let {req, body, streamingOutput, followRedirect, currentRedirect, maxRedirects} = opts;
+    let current = 0;
+
+    currentRedirect = currentRedirect || 0;
+    maxRedirects = maxRedirects || 10;
+
+    // Since streams aren't seekable in node, we want to ensure that
+    // whenever 
+    if (body && body instanceof stream.Readable || body && typeof body.pipe === 'function') {
+      debug('WARNING: instances of stream.Readable are not retryable or redirectable');
```

Comment:
Hmm, what if instead of streams we support file objects? In golang there is an extension to streams that implement `tell()` and `seek()`, but node doesn't seem to have that.
However, if we support file descriptors we can supply supply `fd` to `createReadStream` and we get both streaming and retries..
See: https://nodejs.org/api/fs.html#fs_fs_createreadstream_path_options

Hmm, I have never tried it.. just thinking... I don't know if we can create two concurrent read streams from one `fd`.. but otherwise I think `fd`s can be cloned.. as far as making requests is concerned we don't need concurrent stream though that's only when uploading in parallel.

The thinking is that if we can avoid support streams, that'll prevent footguns like trying to upload a live log :)